packages/design-toolkit/src/hooks/coordinate-field/use-coordinate-field-state.ts

```diff
@@ -0,0 +1,253 @@
+  convertDisplaySegmentsToDD,
+  validateCoordinateSegments,
+} from '../../components/coordinate-field/coordinate-utils';
+import { getSegmentConfigs } from '../../components/coordinate-field/segment-configs';
+import type {
+  CoordinateSystem,
+  CoordinateValue,
+  SegmentConfig,
+} from '../../components/coordinate-field/types';
+
+export interface UseCoordinateFieldStateOptions {
+  value?: CoordinateValue | null;
+  defaultValue?: CoordinateValue;
+  format: CoordinateSystem;
+  onChange?:
+    | Dispatch<SetStateAction<CoordinateValue | null>>
+    | ((value: CoordinateValue | null) => void);
+  onError?: (message: string, context?: Record<string, unknown>) => void;
+  registerTimeout?: (timeoutId: NodeJS.Timeout) => void;
+}
+
+export interface UseCoordinateFieldStateResult {
+  currentValue: CoordinateValue | null;
+  segmentValues: string[];
+  validationErrors: string[];
+  segmentConfigs: SegmentConfig[];
+  editableSegmentConfigs: SegmentConfig[];
+  handleSegmentChange: (index: number, newValue: string) => void;
+  setSegmentValues: (values: string[]) => void;
+  setValidationErrors: (errors: string[]) => void;
+  effectiveErrorMessage: string | null;
+  applyPastedCoordinate: (pastedValue: CoordinateValue) => void;
+  flushPendingValidation: () => void;
+}
+
+export function useCoordinateFieldState({
+  value,
+  defaultValue,
+  format,
+  onChange,
+  onError,
+  registerTimeout,
+}: UseCoordinateFieldStateOptions): UseCoordinateFieldStateResult {
+  const [internalValue, setInternalValue] = useState<CoordinateValue | null>(
+    defaultValue || null,
+  );
+  const [validationErrors, setValidationErrors] = useState<string[]>([]);
+  const validationTimeoutRef = useRef<NodeJS.Timeout | null>(null);
+
+  const currentValue = value !== undefined ? value : internalValue;
+
+  const handleChange = (newValue: CoordinateValue | null) => {
+    if (value === undefined) {
+      setInternalValue(newValue);
+    }
+    onChange?.(newValue);
+  };
+
+  const segmentConfigs = useMemo(() => getSegmentConfigs(format), [format]);
+
+  const editableSegmentConfigs = useMemo(
+    () => segmentConfigs.filter((config) => config.type !== 'literal'),
+    [segmentConfigs],
+  );
+
+  const [segmentValues, setSegmentValues] = useState<string[]>(() => {
+    const initialValue = value !== undefined ? value : defaultValue;
+    if (!initialValue) {
+      return new Array(editableSegmentConfigs.length).fill('');
+    }
+
+    const segments = convertDDToDisplaySegments(initialValue, format);
+    if (!segments) {
+      return new Array(editableSegmentConfigs.length).fill('');
+    }
+
+    return segments;
+  });
+
+  const convertValueToSegmentsOrClear = useCallback(
+    (valueToConvert: CoordinateValue | null) => {
+      // Clear any pending validation when value changes
+      if (validationTimeoutRef.current) {
+        clearTimeout(validationTimeoutRef.current);
+        validationTimeoutRef.current = null;
+      }
```

Comment:
you repeat this check in multiple places - could extract to a helper function and call:

```
const clearValidationTimeout = () => {
  if (validationTimeoutRef.current) {
    clearTimeout(validationTimeoutRef.current);
    validationTimeoutRef.current = null;
  }
};
```